src/sources/pulsar.rs

```diff
@@ -0,0 +1,624 @@
+//! `Pulsar` source.
+//! Accepts log events streamed from [`Apache Pulsar`][pulsar].
+//!
+//! [pulsar]: https://pulsar.apache.org/
+use chrono::TimeZone;
+use futures_util::StreamExt;
+use pulsar::{
+    authentication::oauth2::{OAuth2Authentication, OAuth2Params},
+    consumer::Message,
+    error::AuthenticationError,
+    message::proto::MessageIdData,
+    Authentication, Consumer, Pulsar, SubType, TokioExecutor,
+};
+use tokio_util::codec::FramedRead;
+
+use vector_lib::{
+    codecs::{
+        decoding::{DeserializerConfig, FramingConfig},
+        StreamDecodingError,
+    },
+    config::{LegacyKey, LogNamespace, SourceAcknowledgementsConfig, SourceOutput},
+    configurable::configurable_component,
+    event::Event,
+    finalization::BatchStatus,
+    finalizer::OrderedFinalizer,
+    internal_event::{
+        ByteSize, BytesReceived, CountByteSize, EventsReceived, InternalEventHandle, Protocol,
+        Registered,
+    },
+    sensitive_string::SensitiveString,
+    shutdown::ShutdownSignal,
+    ByteSizeOf,
+};
+use vrl::{owned_value_path, path, value::Kind};
+
+use crate::{
+    codecs::{Decoder, DecodingConfig},
+    config::{SourceConfig, SourceContext},
+    event::BatchNotifier,
+    internal_events::{
+        PulsarErrorEvent, PulsarErrorEventData, PulsarErrorEventType, StreamClosedError,
+    },
+    serde::{bool_or_struct, default_decoding, default_framing_message_based},
+    SourceSender,
+};
+
+/// Configuration for the `pulsar` source.
+#[configurable_component(source("pulsar", "Collect logs from Apache Pulsar."))]
+#[derive(Clone, Debug, Derivative)]
+#[derivative(Default)]
+#[serde(deny_unknown_fields)]
+pub struct PulsarSourceConfig {
+    /// The endpoint to which the Pulsar client should connect to.
+    #[configurable(metadata(docs::examples = "pulsar://127.0.0.1:6650"))]
+    #[serde(alias = "address")]
+    endpoint: String,
+
+    /// The Pulsar topic names to read events from.
+    #[configurable(metadata(docs::examples = "[persistent://public/default/my-topic]"))]
+    topics: Vec<String>,
+
+    /// The Pulsar consumer name.
+    #[configurable(metadata(docs::examples = "consumer-name"))]
+    consumer_name: Option<String>,
+
+    /// The Pulsar subscription name.
+    #[configurable(metadata(docs::examples = "subscription_name"))]
+    subscription_name: Option<String>,
+
+    /// Priority level for a consumer to which a broker gives more priority while dispatching messages in Shared subscription type.
```

Comment:
This first sentence is a little unclear and repetitive given the remaining description. I would suggest keeping the first line simple.
```suggestion
    /// The consumer's priority level.
```